pkg/controllers/management/auth/globalroles/enqueue_test.go

```diff
@@ -911,3 +914,249 @@ func Test_namespaceEnqueueGR(t *testing.T) {
 		})
 	}
 }
+func TestClusterRoleEnqueueGR(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	tests := map[string]struct {
+		obj              runtime.Object
+		grCache          func() mgmtv3.GlobalRoleCache
+		wantKeys         []relatedresource.Key
+		wantErrorMessage string
+	}{
+		"enqueue grb if cr contains the label authz.management.cattle.io/grb-owner": {
+			obj: &v1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						grOwnerLabel: "gr",
+					},
+				},
+			},
+			grCache: func() mgmtv3.GlobalRoleCache {
+				mock := fake.NewMockNonNamespacedCacheInterface[*v3.GlobalRole](ctrl)
+				mock.EXPECT().GetByIndex(grSafeConcatIndex, "gr").Return([]*v3.GlobalRole{{
+					ObjectMeta: metav1.ObjectMeta{Name: "gr"},
+				}}, nil)
+				return mock
+			},
+			wantKeys: []relatedresource.Key{
+				{
+					Name: "gr",
+				},
+			},
+		},
+		"don't enqueue grb if cr doesn't contain the label authz.management.cattle.io/grb-owner": {
+			obj: &v1.ClusterRole{},
+			grCache: func() mgmtv3.GlobalRoleCache {
+				return fake.NewMockNonNamespacedCacheInterface[*v3.GlobalRole](ctrl)
+			},
+			wantKeys: nil,
+		},
+		"nil obj": {
+			obj: nil,
+			grCache: func() mgmtv3.GlobalRoleCache {
+				return fake.NewMockNonNamespacedCacheInterface[*v3.GlobalRole](ctrl)
+			},
+			wantKeys: nil,
+		},
+		"can't cast obj to ClusterRole": {
+			obj: &v1.Role{},
+			grCache: func() mgmtv3.GlobalRoleCache {
+				return fake.NewMockNonNamespacedCacheInterface[*v3.GlobalRole](ctrl)
+			},
+			wantKeys: nil,
+		},
+		"can't get GlobalRole from cache": {
+			obj: &v1.ClusterRole{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						grOwnerLabel: "gr",
+					},
+				},
+			},
+			grCache: func() mgmtv3.GlobalRoleCache {
+				mock := fake.NewMockNonNamespacedCacheInterface[*v3.GlobalRole](ctrl)
+				mock.EXPECT().GetByIndex(grSafeConcatIndex, "gr").Return(nil, errors.New("unexpected error"))
+				return mock
+			},
+			wantKeys:         nil,
+			wantErrorMessage: "unable to get GlobalRole gr for RoleBinding : unexpected error",
```

Comment:
I disagree with you here. I believe there is value in verifying the message. The test would pass if we encounter a different error by simply checking if the error is nil. We are already using `assert.EqualError` and `assert.ErrorContains` in several places in the Rancher code.

Reply:
> The test would pass if we encounter a different error by simply checking if the error is nil. 

Right, this is intentional. If I were to change the error here to something different (i.e. to add more context) I don't want the test to fail. 

Think about it this way - let's say that I wrote a function that called this function. Would it be acceptable for me to process the string error message to check for a specific error message? For example, checking for a not found by doing a `strings.HasSuffix(err.Error(), "NotFound")`? If not, then we shouldn't be using this in our tests.

> We are already using assert.EqualError and assert.ErrorContains in several places in the Rancher code.

I don't think this is a sufficient justification for the use here. Rancher has it's share of issues, we need to evaluate it if solutions fit before copying them into net new code.

Reply:
> Right, this is intentional. If I were to change the error here to something different (i.e. to add more context) I don't want the test to fail.

I think it is absolutely fine if the test fails when the error message has changed. The test will clearly show it failed because of the error message, and will show the diff. It should be straightforward to update the test with the new message.

> Think about it this way - let's say that I wrote a function that called this function. Would it be acceptable for me to process the string error message to check for a specific error message? For example, checking for a not found by doing a strings.HasSuffix(err.Error(), "NotFound")? If not, then we shouldn't be using this in our tests.

In that case I would create a custom error type and use type assertion (something like [this](https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go#type-assertions-and-custom-errors)). This would make unit testing easier, but I think it is an overhead we don't really need if we rely on the error messages. We would end up with a lot of custom error types. I'm open to considering this possibility.

> I don't think this is a sufficient justification for the use here. Rancher has it's share of issues, we need to evaluate it if solutions fit before copying them into net new code.

I agree this is not is a sufficient justification for using it. However, I strongly believe it is better to check the error message than just checking if the error is not nil.  To give an example of current tests in Rancher:
- [here](https://github.com/rancher/rancher/blob/v2.8.3/pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go#L414-L415) we just check if we want an error. However, we don't specified which error we are expecting. That means the wrong error might be returned, and the test would pass.
- [here](https://github.com/rancher/rancher/blob/v2.9.0-alpha1/pkg/controllers/dashboard/plugin/fscache_test.go#L237)  we check the error message. That means we verify the right error is returned

Reply:
> In that case I would create a custom error type and use type assertion (something like [this](https://www.digitalocean.com/community/tutorials/creating-custom-errors-in-go#type-assertions-and-custom-errors)).

This is a part of my argument here. If we aren't expecting people to use the string error message for meaningful information, then why are we asserting on it in a test? To put it another way, the string error message isn't a part of the interface - the caller has no guarantee of what (if anything) will be included in that part of the error message.

> However, I strongly believe it is better to check the error message than just checking if the error is not nil.

I disagree with you here.

> - [here](https://github.com/rancher/rancher/blob/v2.8.3/pkg/controllers/management/auth/globalroles/globalrolebinding_handler_test.go#L414-L415) we just check if we want an error. However, we don't specified which error we are expecting. That means the wrong error might be returned, and the test would pass.
> - [here](https://github.com/rancher/rancher/blob/v2.9.0-alpha1/pkg/controllers/dashboard/plugin/fscache_test.go#L237) we check the error message. That means we verify the right error is returned

My biggest problem with these comments is that the "right error" isn't based in the code. The interface of the function provides no guarantees on the type of error returned. "Some error" is equally as valid as the current things that you are asserting on. As you have said above, you don't consider this output consumable for a real function, so it really shouldn't be something that you test.